include/deal.II/integrators/grad_div.h

```diff
@@ -0,0 +1,379 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2010 - 2016 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE at
+// the top level of the deal.II distribution.
+//
+// ---------------------------------------------------------------------
+
+#ifndef dealii__integrators_grad_div_h
+#define dealii__integrators_grad_div_h
+
+
+#include <deal.II/base/config.h>
+#include <deal.II/base/exceptions.h>
+#include <deal.II/base/quadrature.h>
+#include <deal.II/lac/full_matrix.h>
+#include <deal.II/fe/mapping.h>
+#include <deal.II/fe/fe_values.h>
+#include <deal.II/meshworker/dof_info.h>
+
+DEAL_II_NAMESPACE_OPEN
+
+namespace LocalIntegrators
+{
+  /**
+   * Local integrators related to the grad-div operator and its boundary
+   * traces
+   *
+   * @ingroup Integrators
+   * @author Guido Kanschat, Timo Heister
+   * @date 2016
+   */
+  namespace GradDiv
+  {
+    /**
+     * The weak form of the grad-div operator penalizing volume changes
+     * @f[
+     *  \int_Z \nabla\!\cdot\!u \nabla\!\cdot\!v \,dx
+     * @f]
+     *
+     * @author Guido Kanschat
+     * @date 2011
+     */
+    template <int dim>
+    void cell_matrix (
+      FullMatrix<double> &M,
+      const FEValuesBase<dim> &fe,
+      double factor = 1.)
+    {
+      const unsigned int n_dofs = fe.dofs_per_cell;
+
+      AssertDimension(fe.get_fe().n_components(), dim);
+      AssertDimension(M.m(), n_dofs);
+      AssertDimension(M.n(), n_dofs);
+
+      for (unsigned int k=0; k<fe.n_quadrature_points; ++k)
+        {
+          const double dx = factor * fe.JxW(k);
+          for (unsigned int i=0; i<n_dofs; ++i)
+            for (unsigned int j=0; j<n_dofs; ++j)
+              {
+                double dv = 0.;
+                double du = 0.;
+                for (unsigned int d=0; d<dim; ++d)
+                  {
+                    dv += fe.shape_grad_component(i,k,d)[d];
```

Comment:
I would have said that using `shape_grad_component` is more efficient and thus should be preferred in a library function, but when looking at how the `divergence` implementation i `FEValuesExtracts::Vector` is implemented, I'd say that it cannot be much slower, so it might make sense indeed. (There are some checks in there but I don't think they are more expensive than what appears through `dim` times `shape_grad_component`).